pyro/distributions/polynomial.py

```diff
@@ -0,0 +1,139 @@
+    polynomial flow can be scored.
+
+    :param autoregressive_nn: an autoregressive neural network whose forward call returns a tensor of real-valued
+        numbers of size (batch_size, (count_degree+1)*count_sum, input_dim)
+    :type autoregressive_nn: nn.Module
+    :param count_degree: The degree of the polynomial to use for each element-wise transformation.
+    :type count_degree: int
+    :param count_sum: The number of polynomials to sum in each element-wise transformation.
+    :type count_sum: int
+
+    References:
+
+    Sum-of-squares polynomial flow. [arXiv:1905.02325]
+    Priyank Jaini, Kira A. Shelby, Yaoliang Yu
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 1
+    autoregressive = True
+
+    def __init__(self, autoregressive_nn, input_dim, count_degree, count_sum):
+        super(PolynomialFlow, self).__init__(cache_size=1)
+
+        self.arn = autoregressive_nn
+        self.input_dim = input_dim
+        self.count_degree = count_degree
+        self.count_sum = count_sum
+        self._cached_logDetJ = None
+
+        self.c = nn.Parameter(torch.Tensor(input_dim))
+        self.reset_parameters()
+
+        # Vector of powers of input dimension
+        powers = torch.arange(1, count_degree + 2, dtype=torch.get_default_dtype())
+        self.register_buffer('powers', powers)
+
+        # Build mask of constants
+        mask = torch.zeros(count_degree + 1, count_degree + 1)
+        for jdx in range(count_degree + 1):
+            mask.data[jdx] = self.powers + jdx
+        power_mask = mask
+        mask = mask.reciprocal()
+
+        self.register_buffer('power_mask', power_mask)
+        self.register_buffer('mask', mask)
+
+    def reset_parameters(self):
+        stdv = 1. / math.sqrt(self.c.size(0))
+        self.c.data.uniform_(-stdv, stdv)
+
+    def _call(self, x):
+        """
+        :param x: the input into the bijection
+        :type x: torch.Tensor
+
+        Invokes the bijection x=>y; in the prototypical context of a TransformedDistribution `x` is a
+        sample from the base distribution (or the output of a previous flow)
+        """
+        # Calculate the polynomial coefficients
+        # ~ (batch_size, count_sum, count_degree+1, input_dim)
+        A = self.arn(x).view(-1, self.count_sum, self.count_degree + 1, self.input_dim)
+
+        # Take cross product of coefficients across degree dim
+        # ~ (batch_size, count_sum, count_degree+1, count_degree+1, input_dim)
+        coefs = A.unsqueeze(-2) * A.unsqueeze(-3)
+
+        # Calculate output as sum-of-squares polynomial
+        x_view = x.view(-1, 1, 1, self.input_dim)
+        x_pow_matrix = x_view.pow(self.power_mask.unsqueeze(-1)).unsqueeze(-4)
+
+        # NOTE: The view_as is necessary because the batch dimensions were collapsed previously
+        y = self.c + (coefs * x_pow_matrix * self.mask.unsqueeze(-1)).sum((1, 2, 3)).view_as(x)
```

Comment:
btw is there any way to make this more performant? this might be a bit slow since you're realizing these large 4-dimensional tensors before summing out 3 of the dimensions. e.g. can it be done more greedily or with matrix multiplies or einsums?

Reply:
I don't think it's that slow relative to the alternatives as they're all vectorized operations. This wouldn't be a good use of time to look into, the more important question is how can you actually produce a flexible and stable normalizing flow?